### DISCUSSION TITLE
Proposal for a U.S. Mobile Design System

### DISCUSSION THREAD

**Discussion by mheadd** [uswds/uswds#6101] [Show and tell]
# Proposal for a government design system for native mobile

We wanted to share that our company, Ad Hoc, has begun [working on a version of the US Web Design System specifically tailored to native mobile application](https://adhocteam.us/2024/08/13/us-mobile-design-system/) development. While this work is currently both unofficial and experimental, we've been leaning toward calling it the US Mobile Design System (USMDS), in the hope that one day this will be an official part of our design ecosystem. 

## Background

This effort is informed by our years of work building, deploying, and managing (as part of a larger team of vendors and government employees) the [VA Health and Benefits Mobile App](https://mobile.va.gov/app/va-health-and-benefits). One of the challenges we have encountered in this work is the lack of guidance and resources for how to approach design consideration in a native mobile context “the right way.” We ended up developing our own approach, which incorporated as much as possible from USWDS. However, we believe there is a need for a more clearly defined and (ideally) officially sanctioned approach that would signal to our team, and others, that they are complying with best practices and standards and promoting uniformity of approach across federal agency mobile applications.

Our hope is that this project will get better with community input and feedback from contributions that will hone and sharpen our thinking, and further define the components and resources that are contained within the USMDS. You can check out what we have done so far [with the USMDS on Figma](https://www.figma.com/community/file/1403074855914670787/u-s-mobile-design-system-usmds).

## What we hope to achieve

Our goal is to help foster the development of both tools and guidance that are useful to teams building native mobile applications for federal agencies that want to leverage the power of the USWDS and to promote best practices in native mobile app development. Ultimately, our hope is that this effort will be recognized as an official implementation or approach under the auspices of the USWDS. We know that agencies are developing, and will continue to develop, native mobile applications. We think this effort could go a long way toward developing a critical resource to support the important work of these teams. 

## How you can contribute to the work

First and foremost, if you are working on a native mobile app for a federal agency we’d love to have your feedback in this discussion, and hear your ideas for improving what we’ve already started. 

- Has your team made any use of USWDS as part of its native mobile app development?
- What have you used in your native mobile app from USWDS and where/how have you used it? What have you learned?
- How have your teams balanced using USWDS with following Apple’s Human Interface Guidelines (HIG) and Google’s Material Design guidelines?
- What ideas do you have for resources, guidance, or artifacts that would be useful to your mobile development team (or other teams)?
- What would help your team in developing native mobile apps that incorporate native mobile best practices and the best parts of the USWDS that can be applied to mobile apps?

We’d love to hear from the USWDS community on how we can make this effort work well for native mobile app development teams across the federal government.

**Comment by TimRoe** [comment]
Adding onto this discussion to share a relevant experience. 

Beginning in July 2023, our company, Ad Hoc, performing work for the VA started creating a native mobile design system to serve as a basis for future mobile apps. This was tied to the success of the VA Health and Benefits app ([iOS](https://apps.apple.com/us/app/va-health-and-benefits/id1559609596), [Android](https://play.google.com/store/apps/details?id=gov.va.mobileapp&hl=en_US)), consideration VA offers numerous mobile apps that are web wrappers with mixed reception, an ever growing portion of VA's traffic (like digital traffic generally) has been coming from mobile devices, and that native mobile apps can offer improved device integration (e.g. using biometrics, camera, more intuitive screen reader/accessibility features, offline availability, etc.). 

This work has leaned heavily on USWDS as well as the [VA Design System (VADS)](https://design.va.gov/), but has encountered additional challenges such as: many websites (including VA's) do not support light/dark mode where it's expected for mobile apps; lacking as well defined rules around accessibility and spacing/typography/other atomic tokens that compensate for mobile form factor with less screen real estate; that HTML/CSS paired with modern browsers inherently support a set of default features not necessarily present in native mobile that require additional engineering/components to implement; and that some components routinely used in mobile apps are a product of the different form factor and simply do not exist in web browsers.

Lastly, before getting into more details, one challenge is simply that most enterprises/organizations, with the exception of Apple and Google that make the mobile OSes themselves, do not have a broad enough need to create a native mobile design system because they often only need one app or a few compared to dozens, hundreds, or more web pages so there are simply far fewer existing resources demonstrating what a native mobile design system should be like relative to web design systems. The US government as a whole does not have such a luxury where one app could satisfy needs; many individual agencies have broad enough scope to warrant multiple apps.

So, to get into firm details:
- [Here is the design system section of the VA: Health and Benefits app documentation site](https://department-of-veterans-affairs.github.io/va-mobile-app/design/Intro)
   - Specifically, I'd call attention to the [engineer overview](https://department-of-veterans-affairs.github.io/va-mobile-app/design/About/For%20engineers/overview) for high level technical explanation; basic design principles largely follow VADS linked above where possible
   - The full technical details can be found [in the public project repo](https://github.com/department-of-veterans-affairs/va-mobile-library)
- There is also [a Storybook available](https://department-of-veterans-affairs.github.io/va-mobile-library/?path=/docs/alert--docs) demonstrating the components and tokens created so far
   - Rendered for web using the `react-native-web` NPM package
   - Can also be run directly on mobile devices using the Expo Go app with some extra steps described [here](https://github.com/department-of-veterans-affairs/va-mobile-library/tree/main/packages/components#for-contributors) (should be able to use GitHub Codespaces to save local install effort)

To close, I'd like to note how helpful the existence of the proposed USMDS would be for the VA as, as far as I'm aware, the first US government agency to wade into creating a native mobile design system. There are doubtlessly other native mobile apps being developed within the US government and doubtlessly will be more in the future as traffic continues shifting from more traditional computers/laptops towards mobile phones and tablets. In the same way that USWDS has provided a standardized template/branding for the US government websites, it would be invaluable to have the same for US government native mobile apps.

**Comment by mheadd** [comment]
Software company Blen has released a [React Native implementation of the U.S. Web Design System](https://github.com/blencorp/react-native-usmds), bringing USWDS to mobile applications.

**Reply by caseywatts** [reply]
This looks great, thank you for sharing!

If you (or anyone) has a chance, I think it would be great to have this on the [Implementations page](https://designsystem.digital.gov/documentation/implementations/). Here is [the source file to add this to](https://github.com/uswds/uswds-site/blob/main/_data/implementations.yml)).

(I'm not on the USWDS team, I'm just an enthusiastic community member who refers people/organizations to this implementations page a lot!)

**Reply by mheadd** [reply]
Great idea! I'll pass it along.
